src/Core/IdentityServer/ResourceOwnerPasswordValidator.cs

```diff
@@ -60,25 +59,36 @@ public async Task ValidateAsync(ResourceOwnerPasswordValidationContext context)
             //    return;
             //}
 
-            if (_captchaValidationService.ServiceEnabled && _currentContext.IsBot)
+            string bypassToken = null;
+            if (_captchaValidationService.ServiceEnabled && (_currentContext.IsBot || _captchaValidationService.RequireCaptcha))
             {
-                var captchaResponse = context.Request.Raw["CaptchaResponse"]?.ToString();
+                var user = await _userManager.FindByEmailAsync(context.UserName.ToLowerInvariant());
+                var captchaResponse = context.Request.Raw["captchaResponse"]?.ToString();
+
                 if (string.IsNullOrWhiteSpace(captchaResponse))
                 {
-                    context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, "Captcha required.");
+                    context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, "Captcha required.",
+                        new Dictionary<string, object> {
+                            { "HCaptcha_SiteKey", _captchaValidationService.SiteKey },
+                        });
                     return;
                 }
 
-                var captchaValid = await _captchaValidationService.ValidateCaptchaResponseAsync(captchaResponse,
-                    _currentContext.IpAddress);
+                var captchaValid = _captchaValidationService.ValidateCaptchaBypassToken(captchaResponse, user) ||
```

Comment:
Perhaps you could prefix the token with a string so you know it is a bypass token vs a captcha response? That way you could return early instead of doing cryptographic computation each time?